ls1intum/Artemis
# Fix build log bug for participations with failed builds

Fixes a rather critical, but random bug when saving build logs to the database in the Bamboo setup:

```
Error when creating build result from Bamboo notification: null index column for collection: de.tum.in.www1.artemis.domain.ProgrammingSubmission.buildLogEntries; nested exception is org.hibernate.HibernateException: null index column for collection: de.tum.in.www1.artemis.domain.ProgrammingSubmission.buildLogEntries
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]: org.springframework.orm.jpa.JpaSystemException: null index column for collection: de.tum.in.www1.artemis.domain.ProgrammingSubmission.buildLogEntries; nested exception is org.hibernate.HibernateException: null index column for collection: de.tum.in.www1.artemis.domain.ProgrammingSubmission.buildLogEntries
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:331)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:233)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:551)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:242)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:152)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:174)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at com.sun.proxy.$Proxy253.save(Unknown Source)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at de.tum.in.www1.artemis.service.connectors.bamboo.BambooService.onBuildCompleted(BambooService.java:590)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at de.tum.in.www1.artemis.service.ProgrammingExerciseGradingService.processNewProgrammingExerciseResult(ProgrammingExerciseGradingService.java:93)
Jan 04 12:39:09 vmbruegge60 artemis.sh[73784]:         at de.tum.in.www1.artemis.web.rest.ResultResource.notifyNewProgrammingExerciseResult(ResultResource.java:142)
```

It also seems that retrieving build logs in the Jenkins setup randomly fails. In this setup, Artemis still retrieves the build logs from Jenkins every time the user access them. We should also save them in the Artemis database to avoid asking Jenkins all the time.

One additional task for a follow-up PR:
- Update the https://github.com/ls1intum/jenkins-server-notification-plugin so that it automatically send the build logs to Artemis when the build is finished. 


### Checklist

### Motivation and Context



### Description


### Steps for Testing


1. Create a programming exercise on TS1 and TS2 with different configurations (different programming languages, SCA enabled / disabled, etc.)
2. Solve a test case and push
3. Add a compile failure into the code, commit + push or submit the changes.
4. Check that the build log is shown when you click on the result (on all different pages)
5. Check that the build log is automatically shown in the online code editor when you are currently on this page (when the result arrives) --> no additional user input should be needed to see the compile errors in the build log

### Test Coverage
* JenkinsService 46% -> 64%
